circuit-construction/benches/example_proof.rs

```diff
@@ -0,0 +1,128 @@
+use ark_ec::{AffineCurve, ProjectiveCurve};
+use ark_ff::{FftField, PrimeField, UniformRand};
+use ark_poly::{EvaluationDomain, Radix2EvaluationDomain as D};
+use circuit_construction::*;
+use commitment_dlog::{
+    commitment::{CommitmentCurve, PolyComm},
+    srs::{endos, SRS},
+};
+use groupmap::GroupMap;
+use kimchi::verifier::verify;
+use mina_curves::pasta::{
+    fp::Fp,
+    fq::Fq,
+    pallas::{Affine as Other, PallasParameters},
+    vesta::{Affine, VestaParameters},
+};
+use o1_utils::types::fields::*;
+use oracle::{
+    constants::*,
+    poseidon::{ArithmeticSponge, Sponge},
+    sponge::{DefaultFqSponge, DefaultFrSponge},
+};
+use std::sync::Arc;
+
+type SpongeQ = DefaultFqSponge<VestaParameters, PlonkSpongeConstantsKimchi>;
+type SpongeR = DefaultFrSponge<Fp, PlonkSpongeConstantsKimchi>;
+
+type PSpongeQ = DefaultFqSponge<PallasParameters, PlonkSpongeConstantsKimchi>;
+type PSpongeR = DefaultFrSponge<Fq, PlonkSpongeConstantsKimchi>;
+
+pub struct Witness<G: AffineCurve> {
+    pub s: ScalarField<G>,
+    pub preimage: G::BaseField,
+}
+
+// Prove knowledge of discrete log and poseidon preimage of a hash
+pub fn circuit<
+    F: PrimeField + FftField,
+    G: AffineCurve<BaseField = F> + CoordinateCurve,
+    Sys: Cs<F>,
+>(
+    constants: &Constants<F>,
+    // The witness
+    witness: &Option<Witness<G>>,
+    sys: &mut Sys,
+    public_input: Vec<Var<F>>,
+) {
+    let zero = sys.constant(F::zero());
+
+    let constant_curve_pt = |sys: &mut Sys, (x, y)| {
+        let x = sys.constant(x);
+        let y = sys.constant(y);
+        (x, y)
+    };
+
+    let base = constant_curve_pt(sys, G::prime_subgroup_generator().to_coords().unwrap());
+    let scalar = sys.scalar(G::ScalarField::size_in_bits(), || {
+        witness.as_ref().unwrap().s
+    });
+    let actual = sys.scalar_mul(zero, base, scalar);
+
+    let preimage = sys.var(|| witness.as_ref().unwrap().preimage);
+    let actual_hash = sys.poseidon(constants, vec![preimage, zero, zero])[0];
+
+    sys.assert_eq(actual.0, public_input[0]);
+    sys.assert_eq(actual.1, public_input[1]);
+    sys.assert_eq(actual_hash, public_input[2]);
+
+    sys.zk()
```

Comment:
I don't think this should be part of the circuit :o and we should continue to do it transparently